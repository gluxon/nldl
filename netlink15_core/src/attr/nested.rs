use super::NestedAttributesDeserializeError;
use super::NetlinkAttributeDeserializable;
use super::NetlinkAttributeSerializable;
use crate::message::NetlinkPayloadRequest;
use crate::message::NetlinkPayloadResponse;

/// Wraps zero or more child netlink attributes. The Netlink attribute type ID
/// (`nla_type`) for this container will always be `0`.
#[derive(Debug, PartialEq)]
pub struct Nested<T>(pub Vec<T>);

pub const NESTED_ATTR_NLA_TYPE: u16 = 0;

impl<T: NetlinkAttributeSerializable> NetlinkAttributeSerializable for Nested<T> {
    fn get_type(&self) -> u16 {
        NESTED_ATTR_NLA_TYPE
    }

    fn serialize_payload(&self, buf: &mut Vec<u8>) {
        NetlinkPayloadRequest::serialize(&self.0, buf);
    }
}

impl<T: NetlinkAttributeDeserializable> NetlinkAttributeDeserializable for Nested<T> {
    type Error = NestedAttributesDeserializeError<T>;

    fn deserialize(_ty: u16, payload: &[u8]) -> Result<Self, Self::Error> {
        let attributes: Vec<T> = NetlinkPayloadResponse::deserialize(payload)?;
        Ok(Self(attributes))
    }
}
