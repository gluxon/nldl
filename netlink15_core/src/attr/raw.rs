use crate::utils::nla_get_u16;
use crate::utils::ParseNlaIntError;
use std::{convert::TryFrom, mem::size_of};

/// Assists parsing within this crate. Not meant to be used outside the
/// netlink15_core. Instead implementations of
/// `NetlinkAttributeSerializable`/`NetlinkAttributeDeserializable` should be
/// used.
pub(crate) struct RawNetlinkAttribute<'a> {
    pub len: u16,
    pub ty: u16,
    pub payload: &'a [u8],
}

impl<'a> TryFrom<&'a [u8]> for RawNetlinkAttribute<'a> {
    type Error = ParseNlaIntError;

    /// Reads the first 2 bytes from a buffer to determine the Netlink
    /// Attribute's length. The length bytes are interpreted with the host's
    /// natural endianness. This method will panic if the buffer is shorter than
    /// expected.
    ///
    /// It's acceptable to pass a longer buffer than necessary. The remaining
    /// bytes beyond the retrieved length will be ignored.
    // TODO: Make this not panic if the buffer is shorter than the expected
    // payload.
    fn try_from(buf: &'a [u8]) -> Result<Self, Self::Error> {
        let (header_bytes, remaining) = buf.split_at(size_of::<libc::nlattr>());

        let len = nla_get_u16(&header_bytes[0..size_of::<u16>()])?;
        let ty = nla_get_u16(&header_bytes[size_of::<u16>()..2 * size_of::<u16>()])?;
        let payload = {
            let payload_len = usize::from(len) - size_of::<libc::nlattr>();
            &remaining[..payload_len]
        };

        Ok(Self { len, ty, payload })
    }
}
