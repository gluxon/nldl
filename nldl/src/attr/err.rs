#[derive(Debug, thiserror::Error)]
#[error("netlink attribute deserialization failed")]
pub struct DeserializeError {
    pub attribute_struct_name: &'static str,

    /// The Netlink attribute type ID deserialization failed on.
    pub nla_type_id: u16,

    #[source]
    pub source: Box<dyn std::error::Error + Send + Sync>,
}
