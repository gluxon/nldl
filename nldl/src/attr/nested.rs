use super::Deserialize;
use super::DeserializeError;
use super::Serialize;
use crate::message::NetlinkPayloadRequest;
use crate::message::NetlinkPayloadResponse;

/// Wraps zero or more child netlink attributes. The Netlink attribute type ID
/// (`nla_type`) for this container will always be `0`.
#[derive(Debug, PartialEq)]
pub struct Nested<T>(pub Vec<T>);

pub const NESTED_ATTR_NLA_TYPE: u16 = 0;

impl<T: Serialize> Serialize for Nested<T> {
    fn get_type(&self) -> u16 {
        NESTED_ATTR_NLA_TYPE
    }

    fn serialize_payload(&self, buf: &mut Vec<u8>) {
        NetlinkPayloadRequest::serialize(&self.0, buf);
    }
}

impl<T: Deserialize> Deserialize for Nested<T> {
    fn deserialize(ty: u16, payload: &[u8]) -> Result<Self, DeserializeError> {
        let attributes: Vec<T> =
            NetlinkPayloadResponse::deserialize(payload).map_err(|err| DeserializeError {
                attribute_struct_name: "Nested",
                nla_type_id: ty,
                source: Box::new(err),
            })?;

        Ok(Self(attributes))
    }
}
