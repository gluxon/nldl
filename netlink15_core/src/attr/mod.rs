use super::linux::nlmsg_align;
use super::message::{NetlinkPayloadRequest, NetlinkPayloadResponse};
use super::utils::ParseNlaIntError;
use super::write_to_buf_with_prefixed_u16_len;
use std::{convert::TryFrom, fmt::Debug};

mod raw;
use raw::RawNetlinkAttribute;

pub trait NetlinkAttributeSerializable {
    fn get_type(&self) -> u16;
    fn serialize_payload(&self, buf: &mut Vec<u8>);
}

pub trait NetlinkAttributeDeserializable: Debug + Sized + PartialEq {
    type Error: Debug + std::error::Error;
    fn deserialize(ty: u16, payload: &[u8]) -> Result<Self, Self::Error>;
}

impl<T: NetlinkAttributeSerializable> NetlinkPayloadRequest for T {
    fn serialize(&self, buf: &mut Vec<u8>) {
        write_to_buf_with_prefixed_u16_len(buf, |buf| {
            buf.extend_from_slice(&self.get_type().to_ne_bytes()[..]);
            self.serialize_payload(buf);
        });
    }
}

impl<T: NetlinkAttributeSerializable> NetlinkPayloadRequest for Vec<T> {
    fn serialize(&self, buf: &mut Vec<u8>) {
        for attr in self {
            attr.serialize(buf);
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum NestedAttributesDeserializeError<T: NetlinkAttributeDeserializable> {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),

    // There's a cryptic compiler error message when #[error(transparent)] is
    // set on the generic below.
    #[error("{0}")]
    ChildAttributeDeserializeError(T::Error),
}

impl<T: NetlinkAttributeDeserializable> NetlinkPayloadResponse for Vec<T> {
    type Error = NestedAttributesDeserializeError<T>;

    fn deserialize(buf: &[u8]) -> Result<Self, Self::Error> {
        let mut attrs = vec![];
        let mut view = &buf[..];

        while !view.is_empty() {
            let raw = RawNetlinkAttribute::try_from(view)?;
            let RawNetlinkAttribute { len, ty, payload } = raw;

            view = &view[nlmsg_align(usize::from(len))..];

            let attr = T::deserialize(ty, payload)
                .map_err(NestedAttributesDeserializeError::ChildAttributeDeserializeError)?;
            attrs.push(attr);
        }

        Ok(attrs)
    }
}

#[derive(Debug, PartialEq)]
pub struct Nested<T>(pub Vec<T>);

impl<T: NetlinkAttributeDeserializable> NetlinkAttributeDeserializable for Nested<T> {
    type Error = NestedAttributesDeserializeError<T>;

    fn deserialize(_ty: u16, payload: &[u8]) -> Result<Self, Self::Error> {
        let attributes: Vec<T> = NetlinkPayloadResponse::deserialize(payload)?;
        Ok(Self(attributes))
    }
}

#[derive(Debug, PartialEq)]
pub struct UnknownAttribute {
    pub ty: u16,
    pub payload: Vec<u8>,
}

impl NetlinkAttributeSerializable for UnknownAttribute {
    fn get_type(&self) -> u16 {
        self.ty
    }

    fn serialize_payload(&self, buf: &mut Vec<u8>) {
        buf.extend_from_slice(&self.payload[..])
    }
}

impl NetlinkAttributeDeserializable for UnknownAttribute {
    type Error = std::convert::Infallible;

    fn deserialize(ty: u16, payload: &[u8]) -> Result<Self, Self::Error> {
        Ok(Self {
            ty,
            payload: Vec::from(payload),
        })
    }
}
