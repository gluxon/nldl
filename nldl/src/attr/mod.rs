use super::linux::nlmsg_align;
use super::message::NetlinkPayloadRequest;
use super::message::NetlinkPayloadResponse;
use super::write_to_buf_with_prefixed_u16_len;
use raw::RawNetlinkAttribute;
use std::fmt::Debug;
pub use unknown::UnknownAttribute;

mod nested;
mod raw;
mod unknown;

pub use nested::Nested;
pub use raw::ParseRawNetlinkAttributeError;

pub trait Serialize {
    fn get_type(&self) -> u16;
    fn serialize_payload(&self, buf: &mut Vec<u8>);
}

pub trait Deserialize: Debug + Sized + PartialEq {
    type Error: Debug + std::error::Error;
    fn deserialize(ty: u16, payload: &[u8]) -> Result<Self, Self::Error>;
}

impl<T: Serialize> NetlinkPayloadRequest for T {
    fn serialize(&self, buf: &mut Vec<u8>) {
        write_to_buf_with_prefixed_u16_len(buf, |buf| {
            buf.extend_from_slice(&self.get_type().to_ne_bytes()[..]);
            self.serialize_payload(buf);
        });
    }
}

impl<T: Deserialize> NetlinkPayloadResponse for T {
    type Error = ParseNetlinkAttributeFromBufferError<T>;

    fn deserialize(buf: &[u8]) -> Result<Self, Self::Error> {
        let raw = RawNetlinkAttribute::try_from(buf)?;
        Deserialize::deserialize(raw.ty, raw.payload)
            .map_err(ParseNetlinkAttributeFromBufferError::AttributeDeserializeError)
    }
}

impl<T: Serialize> NetlinkPayloadRequest for Vec<T> {
    fn serialize(&self, buf: &mut Vec<u8>) {
        for attr in self {
            attr.serialize(buf);
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum ParseNetlinkAttributeFromBufferError<T: Deserialize> {
    #[error(
        "An error occurred partitioning a buffer into netlink attribute
    #(len, type, payload) fields: {0}"
    )]
    PartitionBufferError(#[from] ParseRawNetlinkAttributeError),

    // There's a cryptic compiler error message when #[error(transparent)] is
    // set on the generic below.
    #[error("{0}")]
    AttributeDeserializeError(T::Error),
}

impl<T: Deserialize> NetlinkPayloadResponse for Vec<T> {
    type Error = ParseNetlinkAttributeFromBufferError<T>;

    fn deserialize(buf: &[u8]) -> Result<Self, Self::Error> {
        let mut attrs = vec![];
        let mut view = buf;

        while !view.is_empty() {
            let raw = RawNetlinkAttribute::try_from(view)?;
            let RawNetlinkAttribute { len, ty, payload } = raw;

            view = &view[nlmsg_align(usize::from(len))..];

            let attr = T::deserialize(ty, payload)
                .map_err(ParseNetlinkAttributeFromBufferError::AttributeDeserializeError)?;
            attrs.push(attr);
        }

        Ok(attrs)
    }
}

#[cfg(feature = "nldl_derive")]
pub use nldl_derive::NetlinkAttributeDeserializable as Deserialize;
#[cfg(feature = "nldl_derive")]
pub use nldl_derive::NetlinkAttributeSerializable as Serialize;
