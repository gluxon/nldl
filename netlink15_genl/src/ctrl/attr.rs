use netlink15_core::attr::{Nested, ParseNetlinkAttributeFromBufferError, UnknownAttribute};
use netlink15_core::utils::NlaGetStringError;
use netlink15_core::utils::ParseNlaIntError;
use netlink15_derive::NetlinkAttributeDeserializable;
use netlink15_derive::NetlinkAttributeSerializable;

const CTRL_ATTR_UNSPEC: u16 = libc::CTRL_ATTR_UNSPEC as u16;
const CTRL_ATTR_FAMILY_ID: u16 = libc::CTRL_ATTR_FAMILY_ID as u16;
const CTRL_ATTR_FAMILY_NAME: u16 = libc::CTRL_ATTR_FAMILY_NAME as u16;
const CTRL_ATTR_VERSION: u16 = libc::CTRL_ATTR_VERSION as u16;
const CTRL_ATTR_HDRSIZE: u16 = libc::CTRL_ATTR_HDRSIZE as u16;
const CTRL_ATTR_MAXATTR: u16 = libc::CTRL_ATTR_MAXATTR as u16;
const CTRL_ATTR_OPS: u16 = libc::CTRL_ATTR_OPS as u16;
const CTRL_ATTR_MCAST_GROUPS: u16 = libc::CTRL_ATTR_MCAST_GROUPS as u16;

// https://www.infradead.org/~tgr/libnl/doc/api/ctrl_8c_source.html#l00043
#[derive(Debug, PartialEq, NetlinkAttributeSerializable, NetlinkAttributeDeserializable)]
#[netlink15(deserialize(error = "ControllerAttributeDeserializeError"))]
pub enum ControllerAttribute {
    #[nla_type(CTRL_ATTR_UNSPEC)]
    Unspec,
    #[nla_type(CTRL_ATTR_FAMILY_ID)]
    FamilyId(u16),
    #[nla_type(CTRL_ATTR_FAMILY_NAME)]
    FamilyName(String),
    #[nla_type(CTRL_ATTR_VERSION)]
    Version(u32),
    #[nla_type(CTRL_ATTR_HDRSIZE)]
    HeaderSize(u32),
    #[nla_type(CTRL_ATTR_MAXATTR)]
    MaxAttr(u32),
    #[nla_type(CTRL_ATTR_OPS)]
    Operations(Nested<ControllerAttributeOperation>),
    #[nla_type(CTRL_ATTR_MCAST_GROUPS)]
    MulticastGroups(Nested<ControllerAttributeMulticastGroup>),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

const CTRL_ATTR_OP_UNSPEC: u16 = libc::CTRL_ATTR_OP_UNSPEC as u16;
const CTRL_ATTR_OP_ID: u16 = libc::CTRL_ATTR_OP_ID as u16;
const CTRL_ATTR_OP_FLAGS: u16 = libc::CTRL_ATTR_OP_FLAGS as u16;

#[derive(Debug, PartialEq, NetlinkAttributeSerializable, NetlinkAttributeDeserializable)]
#[netlink15(deserialize(error = "ParseNlaIntError"))]
pub enum ControllerAttributeOperation {
    #[nla_type(CTRL_ATTR_OP_UNSPEC)]
    Unspec,
    #[nla_type(CTRL_ATTR_OP_ID)]
    Id(u32),
    #[nla_type(CTRL_ATTR_OP_FLAGS)]
    Flags(u32),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

const CTRL_ATTR_MCAST_GRP_UNSPEC: u16 = libc::CTRL_ATTR_MCAST_GRP_UNSPEC as u16;
const CTRL_ATTR_MCAST_GRP_NAME: u16 = libc::CTRL_ATTR_MCAST_GRP_NAME as u16;
const CTRL_ATTR_MCAST_GRP_ID: u16 = libc::CTRL_ATTR_MCAST_GRP_ID as u16;

#[derive(Debug, PartialEq, NetlinkAttributeSerializable, NetlinkAttributeDeserializable)]
#[netlink15(deserialize(error = "ControllerAttributeMulticastGroupDeserializeError"))]
pub enum ControllerAttributeMulticastGroup {
    #[nla_type(CTRL_ATTR_MCAST_GRP_UNSPEC)]
    Unspec,
    #[nla_type(CTRL_ATTR_MCAST_GRP_NAME)]
    Name(String),
    #[nla_type(CTRL_ATTR_MCAST_GRP_ID)]
    Id(u32),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
    #[error(transparent)]
    DeserializeOperationError(
        #[from] ParseNetlinkAttributeFromBufferError<Nested<ControllerAttributeOperation>>,
    ),
    #[error(transparent)]
    DeserializeMulticastGroupError(
        #[from] ParseNetlinkAttributeFromBufferError<Nested<ControllerAttributeMulticastGroup>>,
    ),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeMulticastGroupDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeOperationDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
}
