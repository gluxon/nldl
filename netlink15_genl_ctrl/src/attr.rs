use nldl::attr::Nested;
use nldl::attr::ParseNetlinkAttributeFromBufferError;
use nldl::attr::UnknownAttribute;
use nldl::utils::NlaGetStringError;
use nldl::utils::ParseNlaIntError;

// https://www.infradead.org/~tgr/libnl/doc/api/ctrl_8c_source.html#l00043
#[derive(Debug, PartialEq, nldl::attr::Serialize, nldl::attr::Deserialize)]
#[nldl(deserialize(error = "ControllerAttributeDeserializeError"))]
pub enum ControllerAttribute {
    #[nla_type(libc::CTRL_ATTR_UNSPEC as u16)]
    Unspec,
    #[nla_type(libc::CTRL_ATTR_FAMILY_ID as u16)]
    FamilyId(u16),
    #[nla_type(libc::CTRL_ATTR_FAMILY_NAME as u16)]
    FamilyName(String),
    #[nla_type(libc::CTRL_ATTR_VERSION as u16)]
    Version(u32),
    #[nla_type(libc::CTRL_ATTR_HDRSIZE as u16)]
    HeaderSize(u32),
    #[nla_type(libc::CTRL_ATTR_MAXATTR as u16)]
    MaxAttr(u32),
    #[nla_type(libc::CTRL_ATTR_OPS as u16)]
    Operations(Vec<Nested<ControllerAttributeOperation>>),
    #[nla_type(libc::CTRL_ATTR_MCAST_GROUPS as u16)]
    MulticastGroups(Vec<Nested<ControllerAttributeMulticastGroup>>),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

#[derive(Debug, PartialEq, nldl::attr::Serialize, nldl::attr::Deserialize)]
#[nldl(deserialize(error = "ParseNlaIntError"))]
pub enum ControllerAttributeOperation {
    #[nla_type(libc::CTRL_ATTR_OP_UNSPEC as u16)]
    Unspec,
    #[nla_type(libc::CTRL_ATTR_OP_ID as u16)]
    Id(u32),
    #[nla_type(libc::CTRL_ATTR_OP_FLAGS as u16)]
    Flags(u32),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

#[derive(Debug, PartialEq, nldl::attr::Serialize, nldl::attr::Deserialize)]
#[nldl(deserialize(error = "ControllerAttributeMulticastGroupDeserializeError"))]
pub enum ControllerAttributeMulticastGroup {
    #[nla_type(libc::CTRL_ATTR_MCAST_GRP_UNSPEC as u16)]
    Unspec,
    #[nla_type(libc::CTRL_ATTR_MCAST_GRP_NAME as u16)]
    Name(String),
    #[nla_type(libc::CTRL_ATTR_MCAST_GRP_ID as u16)]
    Id(u32),
    #[nla_type(_)]
    Unknown(UnknownAttribute),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
    #[error(transparent)]
    DeserializeBufferError(#[from] ParseNetlinkAttributeFromBufferError),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeMulticastGroupDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
}

#[derive(thiserror::Error, Debug)]
pub enum ControllerAttributeOperationDeserializeError {
    #[error(transparent)]
    ParseNlaIntError(#[from] ParseNlaIntError),
    #[error(transparent)]
    NlaGetStringError(#[from] NlaGetStringError),
}
