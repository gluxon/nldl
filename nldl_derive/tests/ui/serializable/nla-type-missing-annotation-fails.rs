use nldl::attr::UnknownAttribute;

#[derive(Debug, PartialEq, nldl::attr::Serialize)]
enum ControllerAttributeOperation {
    #[nla_type(0)]
    Unspec,
    #[nla_type(1)]
    Id(u32),
    #[nla_type(2)]
    Flags(u32),
    Unknown(UnknownAttribute)
}

fn main() {}
