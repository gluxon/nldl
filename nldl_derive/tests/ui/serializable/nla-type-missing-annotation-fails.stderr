error: proc-macro derive panicked
 --> tests/ui/serializable/nla-type-missing-annotation-fails.rs:3:28
  |
3 | #[derive(Debug, PartialEq, nldl::attr::Serialize)]
  |                            ^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: Please annotate all enum variants with #[nla_type(..)]. Saw "Unknown" unannotated.
